onnxruntime/python/tools/transformers/models/phi2/README.md

```diff
@@ -0,0 +1,113 @@
+python -m models.phi2.convert_to_onnx -h
+```
+_From wheel:_ \
+Install [ORT nightly package]([https://onnxruntime.ai/docs/install/](https://onnxruntime.ai/docs/install/#inference-install-table-for-all-languages))
+```bash
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx -h
+```
+
+## Export optimized phi2 onnx model for different scenarios
+**Export FP32 ONNX model for Nvidia GPUs** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp32_gpu
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp32_gpu
+```
+\
+**Export FP16 ONNX model for Nvidia GPUs** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp32_gpu
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp32_gpu
+```
+\
+**Export INT4 ONNX model for Nvidia GPUs** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp32_gpu
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp32_gpu
+```
+\
+**Export FP16 ONNX model for Nvidia GPUs with CUDA architecture SM=80~89** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp16_gpu_sm8x
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp16_gpu_sm8x
+```
+\
+**Export INT4 ONNX model for Nvidia GPUs with CUDA architecture SM=80~89** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --int4_gpu_sm8x
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --int4_gpu_sm8x
+```
+\
+**Export FP32 ONNX model for CPU** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp32_cpu
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp32_cpu
+```
+\
+**Export INT4 ONNX model for CPU** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --int4_cpu
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --int4_cpu
+```
+\
+**Export all at once** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp32_cpu --int4_cpu --fp32_gpu --fp16_gpu --int4_gpu --fp16_gpu_sm8x --int4_gpu_sm8x
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp32_cpu --int4_cpu --fp32_gpu --fp16_gpu --int4_gpu --fp16_gpu_sm8x --int4_gpu_sm8x
+```
+## Run example with ORT
+**(e.g) Export FP16 and INT4 ONNX models for Nvidia GPUs with CUDA architecture SM=80~89 and run examples.** \
+_From source:_
+```
+python -m models.phi2.convert_to_onnx --fp16_gpu_sm8x --int4_gpu_sm8x --run_example
+```
+_From wheel:_
+```
+python -m onnxruntime.transformers.models.phi2.convert_to_onnx --fp16_gpu_sm8x --int4_gpu_sm8x --run_example
+```
+The inference example currently supports all models running on CUDA.
+
+## Limitations
+- It's a known issue that symbolic shape inference fails. It can be ignored at the moment as it won't affect the optimized model's inference.

```

Comment:
For the FunctionProtos that get replaced, do you remove them from the list of functions, the list of graph nodes, and the list of value infos? Sometimes, symbolic shape inference fails if all of this information isn't removed.

Removing from list of functions:
https://github.com/microsoft/onnxruntime/blob/3262e8df2f57c649ff0a75ccdc92e07ddaa8e4c4/onnxruntime/python/tools/transformers/onnx_model_bert.py#L123-L140

Removing from list of graph nodes + removing function's value info:
https://github.com/microsoft/onnxruntime/blob/3262e8df2f57c649ff0a75ccdc92e07ddaa8e4c4/onnxruntime/python/tools/transformers/fusion_rotary_attention.py#L1141-L1150

Reply:
good point! let me have a try

Reply:
one of the errors is fixed by adding gqa in symbolic shape infer script. the other one is due to that an invalid onnx(after preprocessing but necessary for fusion) is saved on disk as temp file. since that file will be removed we can ignore it.